pkg/ethereum/nats.go

```diff
@@ -0,0 +1,127 @@
+package ethereum
+
+import (
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/libp2p/go-libp2p/core/peer"
+	eth "github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1"
+)
+
+type PeerDiscoveredEvent struct {
+	ENR        string `json:"enr"`
+	IP         string `json:"ip"`
+	Port       int    `json:"port"`
+	CrawlerID  string `json:"crawler_id"`
+	CrawlerLoc string `json:"crawler_location"`
+}
+
+type MetadataReceivedEvent struct {
+	ENR        string         `json:"enr"`
+	ID         string         `json:"id"`
+	IP         string         `json:"ip"`
+	Port       int            `json:"port"`
+	MetaData   SimpleMetaData `json:"metadata"`
+	CrawlerID  string         `json:"crawler_id"`
+	CrawlerLoc string         `json:"crawler_location"`
+}
+
+type SimpleMetaData struct {
+	SeqNumber uint64
+	Attnets   string
+	Syncnets  []byte
+}
+
+func (n *Node) sendMetadataEvent(ctx context.Context, pid peer.ID, addrInfo peer.AddrInfo, md *eth.MetaDataV1) {
+	addressParts := strings.Split(addrInfo.Addrs[0].String(), "/")
+	ip := addressParts[2]
+	port, err := strconv.Atoi(addressParts[4])
+	if err != nil {
+		n.log.Error().Err(err).Msg("Failed to convert port to int")
+		return
+	}
+	node := n.disc.seenNodes[pid].Node
+
+	metadataEvent := &MetadataReceivedEvent{
+		ENR:        node.String(),
+		ID:         pid.String(),
+		IP:         ip,
+		Port:       port,
+		MetaData:   SimpleMetaData{SeqNumber: md.SeqNumber, Attnets: hex.EncodeToString(md.Attnets), Syncnets: md.Syncnets},
+		CrawlerID:  getCrawlerMachineID(),
+		CrawlerLoc: getCrawlerLocation(),
+	}
+
+	select {
+	case n.metadataEventChan <- metadataEvent:
+		n.log.Debug().Str("peer", pid.String()).Msg("Sent metadata_received event to channel")
+	case <-ctx.Done():
+		n.log.Warn().Msg("Context cancelled before sending metadata_received event to channel")
+	}
+}
+
+func (n *Node) startMetadataPublisher() {
+	go func() {
+		for metadataEvent := range n.metadataEventChan {
+			publishCtx, publishCancel := context.WithTimeout(context.Background(), 3*time.Second)
+			defer publishCancel()
```

Comment:
You should never have `defer` statements inside of a for loop with no clear termination. `defer` only executes when a function exits, which means all of these `defers` will be stored in memory until we get an OOM error! Had this problem multiple times haha

Reply:
Gotcha thanks for the pointer. I resolved this by explicitly calling `publishCancel`. Does that sound good?

Reply:
Perfect!